### ISSUE TITLE
Properly hide visibility of private fields

### ISSUE THREAD

**Issue by xldenis** [creusot-rs/creusot#418]
Private fields should not appear in a type for functions that don't have permission to see them.

**Comment by xldenis** [comment]
We can actually do this very easily at a coarse granularity right now, using [is_visible_locally](https://doc.rust-lang.org/beta/nightly-rustc/rustc_middle/ty/enum.Visibility.html#method.is_visible_locally). This way we could at least restrict the visibility of private fields from external types. This will reduce a lot of garbage currently generated by `Vec` and similar types.

**Comment by jhjourdan** [comment]
I think this is far from easy, because visbility of fields has interaction with opaqueness and partial resolution.

**Comment by xldenis** [comment]
> I think this is far from easy, because visbility of fields has interaction with opaqueness and partial resolution.

Hmm I can see the issue with opaqueness but not with partial resolution, if a field is actually private, you shouldn't ever be in a situation to partially resolve. However, if it also is coming from a crate with no contracts you don't have issues with opacity once again!

**Comment by xldenis** [comment]
The _complete_ solution is tricky, but I think its feasible to get something which works for eg `std`.

**Comment by xldenis** [comment]
> Imagine you have a struct with a public and a private field. You moved the public field. Now, when resolving, you cannot resolve the whole structure (it is not fully initialized), but you can neither resolve the private field (because it is private). What do you do? We probably want to emit an error in this case, but this is a bit tricky!

Yes but we can easily restrict ourselves to the common case of all fields are private. The purpose here would be to avoid spilling all the details of `Vec` as soon as you use it. Same applies for `BTreeMap` for example.

**Comment by jhjourdan** [comment]
Okay, so we would generate an abstract type only if all the fields of all the constructors are private, and there is at least one field in each constructor, and at least one constructor?

**Comment by xldenis** [comment]
Yes, and this should actually have a large impact in practice, I remember prototyping it once, and it cut some of our test files by ~30%. No major impact on performance, but its just cleaner and definitely could have an impact when used with more complex types than `Vec`.
